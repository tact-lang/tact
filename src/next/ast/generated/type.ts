/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as $c from "@/next/ast/common";
import type * as $ from "@/next/ast/type";

export type Signedness = $.Signedness;
export const allSignedness: readonly $.Signedness[] = ["signed", "unsigned"];
export type IFInt = $.IFInt;
export const IFInt = (
    sign: $.Signedness,
    width: number,
    loc: $c.Range,
): $.IFInt =>
    Object.freeze({
        kind: "FInt",
        sign,
        width,
        loc,
    });
export const isIFInt = ($value: IFInt) => $value.kind === "FInt";
export type VarIntWidth = $.VarIntWidth;
export const allVarIntWidth: readonly $.VarIntWidth[] = ["16", "32"];
export type IFVarInt = $.IFVarInt;
export const IFVarInt = (
    sign: $.Signedness,
    width: $.VarIntWidth,
    loc: $c.Range,
): $.IFVarInt =>
    Object.freeze({
        kind: "FVarInt",
        sign,
        width,
        loc,
    });
export const isIFVarInt = ($value: IFVarInt) => $value.kind === "FVarInt";
export type IntFormat = $.IntFormat;
export type TypeInt = $.TypeInt;
export const TypeInt = (format: $.IntFormat, loc: $c.Range): $.TypeInt =>
    Object.freeze({
        kind: "TyInt",
        format,
        loc,
    });
export const isTypeInt = ($value: TypeInt) => $value.kind === "TyInt";
export type SFBits = $.SFBits;
export const SFBits = (bits: number, loc: $c.Range): $.SFBits =>
    Object.freeze({
        kind: "SFBits",
        bits,
        loc,
    });
export const isSFBits = ($value: SFBits) => $value.kind === "SFBits";
export type SFRemaining = $.SFRemaining;
export const SFRemaining = (loc: $c.Range): $.SFRemaining =>
    Object.freeze({
        kind: "SFRemaining",
        loc,
    });
export const isSFRemaining = ($value: SFRemaining) =>
    $value.kind === "SFRemaining";
export type SFDefault = $.SFDefault;
export const SFDefault = (loc: $c.Range): $.SFDefault =>
    Object.freeze({
        kind: "SFDefault",
        loc,
    });
export const isSFDefault = ($value: SFDefault) => $value.kind === "SFDefault";
export type SliceFormat = $.SliceFormat;
export type TypeSlice = $.TypeSlice;
export const TypeSlice = (format: $.SliceFormat, loc: $c.Range): $.TypeSlice =>
    Object.freeze({
        kind: "TySlice",
        format,
        loc,
    });
export const isTypeSlice = ($value: TypeSlice) => $value.kind === "TySlice";
export type RemFormat = $.RemFormat;
export type TypeCell = $.TypeCell;
export const TypeCell = (format: $.RemFormat, loc: $c.Range): $.TypeCell =>
    Object.freeze({
        kind: "TyCell",
        format,
        loc,
    });
export const isTypeCell = ($value: TypeCell) => $value.kind === "TyCell";
export type TypeBuilder = $.TypeBuilder;
export const TypeBuilder = (
    format: $.RemFormat,
    loc: $c.Range,
): $.TypeBuilder =>
    Object.freeze({
        kind: "TyBuilder",
        format,
        loc,
    });
export const isTypeBuilder = ($value: TypeBuilder) =>
    $value.kind === "TyBuilder";
export type TypeUnit = $.TypeUnit;
export const TypeUnit = (loc: $c.Range): $.TypeUnit =>
    Object.freeze({
        kind: "unit_type",
        loc,
    });
export const isTypeUnit = ($value: TypeUnit) => $value.kind === "unit_type";
export type TypeTensor = $.TypeTensor;
export const TypeTensor = (
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeTensor =>
    Object.freeze({
        kind: "tensor_type",
        typeArgs,
        loc,
    });
export const isTypeTensor = ($value: TypeTensor) =>
    $value.kind === "tensor_type";
export type TypeTuple = $.TypeTuple;
export const TypeTuple = (
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeTuple =>
    Object.freeze({
        kind: "tuple_type",
        typeArgs,
        loc,
    });
export const isTypeTuple = ($value: TypeTuple) => $value.kind === "tuple_type";
export type TypeCons = $.TypeCons;
export const TypeCons = (
    name: $c.TypeId,
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeCons =>
    Object.freeze({
        kind: "cons_type",
        name,
        typeArgs,
        loc,
    });
export const isTypeCons = ($value: TypeCons) => $value.kind === "cons_type";
export type Type = $.Type;
export type TypeMap = $.TypeMap;
export const TypeMap = (key: $.Type, value: $.Type, loc: $c.Range): $.TypeMap =>
    Object.freeze({
        kind: "map_type",
        key,
        value,
        loc,
    });
export const isTypeMap = ($value: TypeMap) => $value.kind === "map_type";
