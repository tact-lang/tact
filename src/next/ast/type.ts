import type { Range, TypeId } from "@/next/ast/common";

export type Type =
    | TypeMap
    | TypeCons
    | TypeInt
    | TypeSlice
    | TypeCell
    | TypeBuilder
    | TypeTuple
    | TypeUnit
    | TypeTensor;

export type TypeCons = {
    readonly kind: "cons_type";
    readonly name: TypeId;
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeInt = {
    readonly kind: "TyInt";
    readonly format: IntFormat;
    readonly loc: Range;
};
export type IntFormat = IFInt | IFVarInt;
export type IFInt = {
    readonly kind: "FInt";
    readonly sign: Signedness;
    readonly width: number;
    readonly loc: Range;
};
export type IFVarInt = {
    readonly kind: "FVarInt";
    readonly sign: Signedness;
    readonly width: VarIntWidth;
    readonly loc: Range;
};
export type VarIntWidth = "16" | "32";
export type Signedness = "signed" | "unsigned";

export type TypeSlice = {
    readonly kind: "TySlice";
    readonly format: SliceFormat;
    readonly loc: Range;
};
export type SliceFormat = SFBits | SFRemaining | SFDefault;
export type SFBits = {
    readonly kind: "SFBits";
    readonly bits: number;
    readonly loc: Range;
};
export type SFRemaining = {
    readonly kind: "SFRemaining";
    readonly loc: Range;
};
export type SFDefault = {
    readonly kind: "SFDefault";
    readonly loc: Range;
};

export type TypeCell = {
    readonly kind: "TyCell";
    readonly format: RemFormat;
    readonly loc: Range;
};
export type TypeBuilder = {
    readonly kind: "TyBuilder";
    readonly format: RemFormat;
    readonly loc: Range;
};
export type RemFormat = SFRemaining | SFDefault;

export type TypeTuple = {
    readonly kind: "tuple_type";
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeUnit = {
    readonly kind: "unit_type";
    readonly loc: Range;
};

export type TypeTensor = {
    readonly kind: "tensor_type";
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeMap = {
    readonly kind: "map_type";
    readonly key: Type; // any type except tensor
    readonly value: Type; // any type except tensor
    readonly loc: Range;
};
