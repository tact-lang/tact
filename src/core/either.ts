// left$0 {X:Type} {Y:Type} value:X = Either X Y;
// right$1 {X:Type} {Y:Type} value:Y = Either X Y;

import { uint } from "@/core/numeric";
import type { Tlb } from "@/core/tlb";

export type Either<L, R> = Left<L> | Right<R>
export type Left<L> = { readonly $: 'left', readonly value: L }
export const Left = <L>(value: L): Left<L> => Object.freeze({ $: 'left', value });
export const isLeft = <L>(m: Either<L, unknown>): m is Left<L> => m.$ === 'left';
export type Right<R> = { readonly $: 'right', readonly value: R }
export const Right = <R>(value: R): Right<R> => Object.freeze({ $: 'right', value });
export const isRight = <R>(m: Either<unknown, R>): m is Right<R> => m.$ === 'right';

export const either = <L, R>(left: Tlb<L>, right: Tlb<R>): Tlb<Either<L, R>> => ({
    store: (t, b, p) => {
        if (t.$ === 'left') {
            uint(1).store(0, b, [...p, '$']);
            left.store(t.value, b, [...p, 'value']);
        } else {
            uint(1).store(1, b, [...p, '$']);
            right.store(t.value, b, [...p, 'value']);
        }
    },
    load: (s, p) => {
        const type = uint(1).load(s, [...p, '$']);
        if (type === 0) {
            return Left(left.load(s, [...p, 'value']));
        } else {
            return Right(right.load(s, [...p, 'value']));
        }
    },
});
