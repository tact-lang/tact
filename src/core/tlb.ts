import { entries } from '@/utils/tricks';
import { type Slice, Builder } from '@ton/core'

export type Path = string[]

export class TlbError extends Error {
    constructor(
        public error: unknown,
        public context: Path
    ) {
        super();
    }
}

/**
 * TL-B type
 */
export interface Tlb<T> {
    /**
     * Store `value` into `builder` in context `path`
     */
    store: (value: T, builder: Builder, path: Path) => void;
    /**
     * Load value from `slice` in context `path`
     */
    load: (slice: Slice, path: Path) => T;
}

export type GetTlb<T> = T extends Tlb<infer U> ? U : never

/**
 * Internal wrapper for primitive types
 * Catches errors if load/store failed
 */
export const _primitive = <T>(child: Tlb<T>): Tlb<T> => ({
    store: (t, b, p) => {
        try {
            child.store(t, b, p)
        } catch (e) {
            throw new TlbError(e, p);
        }
    },
    load: (s, p) => {
        try {
            return child.load(s, p);
        } catch (e) {
            throw new TlbError(e, p);
        }
    },
});

/**
 * Non-stored value
 * Always returns `value` when loaded
 * Doesn't have any binary representation
 */
export const pure = <T>(value: T): Tlb<T> => ({
    store: (_t, _b) => {},
    load: (_s) => value,
});

/**
 * Ref to another type
 * Equivalent to `^child` in TL-B
 */
export const ref = <T>(child: Tlb<T>): Tlb<T> => ({
    store: (t, b, p) => {
        const b2 = new Builder();
        const refId = `refs[${b.refs}]`;
        const path = [...p, refId];
        child.store(t, b2, path);
        try {
            b.storeRef(b2.endCell());
        } catch (e) {
            throw new TlbError(e, path);
        }
    },
    load: (s, p) => {
        const refId = `refs[${s.offsetRefs}]`;
        const path = [...p, refId];
        try {
            return child.load(s.loadRef().asSlice(), path);
        } catch (e) {
            throw new TlbError(e, path);
        }
    },
});

/**
 * Sequence of types
 * Equivalent to space-separated list in TL-B:
 * `foo:T bar:U`
 */
export const object = <T extends object>(
    children: { [K in keyof T]: Tlb<T[K]> }
): Tlb<T> => {
    const fields = entries(children);
    return {
        store: (t, b, p) => {
            for (const [k, child] of fields) {
                child.store(t[k], b, [...p, String(k)]);
            }
        },
        load: (s, p) => {
            const result = {} as T;
            for (const [k, child] of fields) {
                result[k] = child.load(s, [...p, String(k)]);
            }
            return result;
        },
    };
};

/**
 * Exact value of type `child`
 * Equivalent to `#1234` in TL-B
 */
export const literal = <const T>(child: Tlb<T>) =>
    <U extends T>(value: U): Tlb<U> => ({
    store: (t, b, p) => {
        if (t !== value) {
            throw new TlbError(new Error('Invalid value'), p);
        }
        child.store(value, b, p);
    },
    load: (s, p) => {
        const readValue = child.load(s, p);
        if (value !== readValue) {
            throw new TlbError(new Error('Invalid value'), p);
        }
        return value;
    },
});