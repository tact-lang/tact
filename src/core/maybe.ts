// nothing$0 {X:Type} = Maybe X;
// just$1 {X:Type} value:X = Maybe X;

import { TlbError, type Path, type Tlb } from "@/core/tlb";
import type { Builder, Slice } from "@ton/core";

export type Maybe<T> = Nothing | Just<T>
export type Nothing = { readonly $: 'nothing' }
export const Nothing: Nothing = Object.freeze({ $: 'nothing' })
export const isNothing = <T>(m: Maybe<T>): m is Nothing => m.$ === 'nothing';
export type Just<T> = { readonly $: 'just', readonly value: T }
export const Just = <T>(value: T): Just<T> => Object.freeze({ $: 'just', value });
export const isJust = <T>(m: Maybe<T>): m is Just<T> => m.$ === 'just';

export const maybe = <T>(type: Tlb<T>): Tlb<Maybe<T>> => {
    const store = (t: Maybe<T>, b: Builder, p: Path) => {
        if (t.$ === 'nothing') {
            b.storeBit(0);
            return;
        }
        try {
            b.storeBit(1);
        } catch (e) {
            throw new TlbError(e, [...p, '$']);
        }
        type.store(t.value, b, [...p, 'value']);
    };
    const load = (s: Slice, p: Path) => {
        const bit = (() => {
            try {
                return s.loadBit();
            } catch (e) {
                throw new TlbError(e, [...p, '$']);
            }
        })();
        if (bit) {
            return Just(type.load(s, [...p, 'value']));
        } else {
            return Nothing;
        }
    };
    return { store, load };
};