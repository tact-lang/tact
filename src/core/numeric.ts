import { _primitive, type Tlb } from "@/core/tlb";

export const bit: Tlb<boolean> = _primitive({
    store: (t, b) => b.storeBit(t),
    load: (s) => s.loadBit(),
});

export const int = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeInt(t, bits),
    load: (s) => s.loadInt(bits),
});

export const uint = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeUint(t, bits),
    load: (s) => s.loadUint(bits),
});

export const intVar = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeVarInt(t, bits),
    load: (s) => s.loadVarInt(bits),
});

export const uintVar = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeVarUint(t, bits),
    load: (s) => s.loadVarUint(bits),
});

export const coins = uintVar(16);

export const intBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeInt(t, bits),
    load: (s) => s.loadIntBig(bits),
});

export const uintBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeUint(t, bits),
    load: (s) => s.loadUintBig(bits),
});

export const intVarBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeVarInt(t, bits),
    load: (s) => s.loadVarIntBig(bits),
});

export const uintVarBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeVarUint(t, bits),
    load: (s) => s.loadVarUintBig(bits),
});