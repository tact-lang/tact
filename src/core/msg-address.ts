// addr_none$00 = MsgAddressExt;
// addr_extern$01 len:(## 9) external_address:(bits len) = MsgAddressExt;
// anycast_info$_ depth:(#<= 30) { depth >= 1 }
//    rewrite_pfx:(bits depth) = Anycast;
// addr_std$10 anycast:(Maybe Anycast) workchain_id:int8 address:bits256  = MsgAddressInt;
// addr_var$11 anycast:(Maybe Anycast) addr_len:(## 9) workchain_id:int32 address:(bits addr_len) = MsgAddressInt;
// _ _:MsgAddressInt = MsgAddress;
// _ _:MsgAddressExt = MsgAddress;

import { int, uint, uintBig } from "@/core/numeric";
import type { Path, Tlb } from "@/core/tlb";
import type { Builder, Slice } from "@ton/core";

export type MsgAddress =
    | AddrNone
    | AddrExtern
    | AddrStd
    | AddrVar
export type AddrNone = {
    readonly $: 'AddrNone'
}
export const AddrNone: AddrNone = Object.freeze({ $: "AddrNone" });
export const isAddrNone = (a: MsgAddress): a is AddrNone => a.$ === 'AddrNone';
export type AddrExtern = {
    readonly $: 'AddrExtern'
    readonly len: number
    readonly address: bigint; // bits len
}
export const AddrExtern = (len: number, address: bigint): AddrExtern => Object.freeze({
    $: 'AddrExtern', address, len,
})
export const isAddrExtern = (a: MsgAddress): a is AddrExtern => a.$ === 'AddrExtern';
export type AddrStd = {
    readonly $: 'AddrStd'
    readonly workchainId: number;
    readonly address: bigint;
}
export const AddrStd = (workchainId: number, address: bigint): AddrStd => Object.freeze({
    $: 'AddrStd', address, workchainId,
})
export const isAddrStd = (a: MsgAddress): a is AddrStd => a.$ === 'AddrStd';
export type AddrVar = {
    readonly $: 'AddrVar'
    readonly len: number // ## 9
    readonly workchainId: number // int32
    readonly address: bigint // bits addr_len
}
export const AddrVar = (len: number, workchainId: number, address: bigint): AddrVar => Object.freeze({
    $: "AddrVar", address, len, workchainId,
});
export const isAddrVar = (a: MsgAddress): a is AddrVar => a.$ === 'AddrVar';

const loadAddress = (s: Slice, p: Path): MsgAddress => {
    const type = uint(2).load(s, [...p, '$']);
    if (type === 0) {
        // addr_none$00
        return AddrNone;
    } else if (type === 1) {
        // addr_extern$01
        // len:(## 9)
        const len = uint(9).load(s, [...p, 'len']);
        // external_address:(bits len)
        const addr = uintBig(len).load(s, [...p, 'address']);
        return AddrExtern(len, addr);
    } else if (type === 2) {
        // addr_std$10
        // anycast:(Maybe Anycast)
        if (uint(1).load(s, [...p, 'anycast']) !== 0) {
            throw Error('Anycast not supported');
        }
        // workchain_id:int8
        const wc = int(8).load(s, [...p, 'workchainId']);
        // address:bits256
        const addr = uintBig(256).load(s, [...p, 'address']);
        return AddrStd(wc, addr);
    } else if (type === 3) {
        // addr_var$11
        // anycast:(Maybe Anycast)
        if (uint(1).load(s, [...p, 'anycast']) !== 0) {
            throw Error('Anycast not supported');
        }
        // addr_len:(## 9)
        const len = uint(9).load(s, [...p, 'len']);
        // workchain_id:int32
        const wc = int(32).load(s, [...p, 'workchainId']);
        // address:(bits addr_len)
        const addr = uintBig(len).load(s, [...p, 'address']);
        return AddrVar(len, wc, addr);
    } else {
        throw Error('Unreachable');
    }
};

const storeAddress = (t: MsgAddress, b: Builder, p: Path) => {
    switch (t.$) {
        case 'AddrNone': {
            // addr_none$00
            uint(2).store(0, b, [...p, '$']);
            return;
        }
        case 'AddrExtern': {
            // addr_extern$01
            uint(2).store(1, b, [...p, '$']);
            // len:(## 9)
            uint(9).store(t.len, b, [...p, 'len']);
            // external_address:(bits len)
            uintBig(t.len).store(t.address, b, [...p, 'address']);
            return;
        }
        case 'AddrStd': {
            // addr_std$10
            uint(2).store(2, b, [...p, '$']);
            // anycast:(Maybe Anycast)
            uint(1).store(0, b, [...p, 'anycast']);
            // workchain_id:int8
            int(8).store(t.workchainId, b, [...p, 'workchainId']);
            // address:bits256
            uintBig(256).store(t.address, b, [...p, 'address']);
            return;
        }
        case 'AddrVar': {
            // addr_var$11
            uint(2).store(3, b, [...p, '$']);
            // anycast:(Maybe Anycast)
            uint(1).store(0, b, [...p, 'anycast']);
            // addr_len:(## 9)
            uint(9).store(t.len, b, [...p, 'len']);
            // workchain_id:int32
            int(32).store(t.workchainId, b, [...p, 'workchainId']);
            // address:(bits addr_len)
            uintBig(t.len).store(t.address, b, [...p, 'address']);
        }
    }
};

export const msgAddress: Tlb<MsgAddress> = ({
    store: storeAddress,
    load: loadAddress,
});