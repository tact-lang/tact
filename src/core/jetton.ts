// InternalMsgBody
//     transfer
//         #0f8a7ea5
//         query_id:uint64
//         amount:(VarUInteger 16)
//         destination:MsgAddress
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     transfer_notification
//         #7362d09c
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         forward_payload:(Either Cell ^Cell)
//     excesses
//         #d53276db
//         query_id:uint64
//     burn
//         #595f07bc
//         query_id:uint64
//         amount:(VarUInteger 16)
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//     internal_transfer
//         #178d4519
//         query_id:uint64
//         amount:(VarUInteger 16)
//         from:MsgAddress
//         response_address:MsgAddress
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     burn_notification
//         #7bdd97de
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         response_destination:MsgAddress

import { either } from "@/core/either";
import { maybe } from "@/core/maybe";
import { msgAddress } from "@/core/msg-address";
import { uint, uintVar } from "@/core/numeric";
import { type GetTlb, literal, object, ref, type Tlb } from "@/core/tlb";

export const transfer = <Custom, Forward>(
    custom: Tlb<Custom>,
    forward: Tlb<Forward>,
) => {
    return object({
        $: literal(uint(32))(0x0f8a7ea5),
        queryId: uint(64),
        amount: uintVar(16),
        destination: msgAddress,
        responseDestination: msgAddress,
        customPayload: maybe(ref(custom)),
        forwardTonAmount: uintVar(16),
        forwardPayload: either(forward, ref(forward)),
    });
};

export type Transfer<T, U> = GetTlb<ReturnType<typeof transfer<T, U>>>
