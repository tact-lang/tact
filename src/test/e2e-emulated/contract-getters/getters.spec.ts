import { beginCell, toNano } from "@ton/core";
import type { SandboxContract, TreasuryContract } from "@ton/sandbox";
import { Blockchain } from "@ton/sandbox";
import { Test, Test_getterMapping } from "./output/getters_Test";
import { Test as Test2 } from "./output/empty-message-getter-parameter_Test";
import "@ton/test-utils";

// disable tests on MacOS
const it = process.platform === "darwin" && process.env.CI ? test.skip : test;

describe("getters", () => {
    let blockchain: Blockchain;
    let treasury: SandboxContract<TreasuryContract>;
    let contract: SandboxContract<Test>;
    let contract2: SandboxContract<Test2>;

    beforeEach(async () => {
        blockchain = await Blockchain.create();
        blockchain.verbosity.print = false;
        treasury = await blockchain.treasury("treasury");

        contract = blockchain.openContract(await Test.fromInit());
        contract2 = blockchain.openContract(await Test2.fromInit());

        const deployResult = await contract.send(
            treasury.getSender(),
            { value: toNano("10") },
            {
                $$type: "Deploy",
                queryId: 0n,
            },
        );

        expect(deployResult.transactions).toHaveTransaction({
            from: treasury.address,
            to: contract.address,
            success: true,
            deploy: true,
        });

        const deployResult2 = await contract2.send(
            treasury.getSender(),
            { value: toNano("10") },
            null,
        );

        expect(deployResult2.transactions).toHaveTransaction({
            from: treasury.address,
            to: contract2.address,
            success: true,
            deploy: true,
        });
    });

    it("should implement getters correctly", async () => {
        // Getter name conflicts
        expect(await contract.getTestGetter()).toBe(1n);
        expect(await contract.gettest_getter()).toBe(2n);
        expect(await contract.getTest_getter()).toBe(3n);

        expect(Test_getterMapping["testGetter"]).toBe("getTestGetter");
        expect(Test_getterMapping["test_getter"]).toBe("gettest_getter");
        expect(Test_getterMapping["Test_getter"]).toBe("getTest_getter");

        // Passing `S` struct to getter
        expect(
            await contract.getStructAsInput({
                $$type: "S",
                a: 1n,
                b: 2n,
            }),
        ).toMatchSnapshot();

        // Returning `self` from getter
        expect(await contract.getContractData()).toMatchSnapshot();

        // Passing `SetIdAndData` message to getter
        expect(
            await contract.getMessageAsInput1({
                $$type: "SetIdAndData",
                id: 42n,
                data: beginCell().endCell(),
            }),
        ).toBe(42n);
        expect(
            await contract.getMessageAsInput2({
                $$type: "SetIdAndData",
                id: 42n,
                data: beginCell().endCell(),
            }),
        ).toMatchSnapshot();

        // Passing `Test` contract data to getter
        expect(
            await contract.getContractAsInput({
                $$type: "Test$Data",
                id: 123n,
                anotherData: beginCell().storeUint(123, 64).endCell(),
            }),
        ).toMatchSnapshot();

        expect(await contract.getMethodIdExpr()).toBe(true);
        expect(await contract.getMethodIdConst()).toBe(2n ** 14n);
        expect(await contract.getMethodIdMin()).toBe(true);
        expect(await contract.getMethodIdMax()).toBe(true);
    });

    it("should take empty message as parameter", async () => {
        const res = await contract2.getFoo({ $$type: "Foo" });
        expect(res).toMatchObject({});
    });
});
