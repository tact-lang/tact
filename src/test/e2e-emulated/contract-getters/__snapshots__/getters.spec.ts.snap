// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getters should implement getters correctly 1`] = `
{
  "$$type": "S",
  "a": 1n,
  "b": 2n,
}
`;

exports[`getters should implement getters correctly 2`] = `
{
  "$$type": "Test$Data",
  "anotherData": x{},
  "id": 0n,
}
`;

exports[`getters should implement getters correctly 3`] = `
{
  "$$type": "SetIdAndData",
  "data": x{},
  "id": 42n,
}
`;

exports[`getters should implement getters correctly 4`] = `
{
  "$$type": "Test$Data",
  "anotherData": x{000000000000007B},
  "id": 123n,
}
`;
